saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationRequestFactory.java

```diff
@@ -58,6 +74,61 @@ public String createAuthenticationRequest(Saml2AuthenticationRequest request) {
 		);
 	}
 
+	@Override
+	public MultiValueMap<String, String> createAuthenticationRequestParameters(Saml2AuthenticationRequest request) {
+		Saml2AuthenticationRequest actual = request;
+		if (request.isDeflate()) {
+			//for deflate, signature comes later
+			actual = builder(request)
+					.credentials(c -> c.clear())
+					.build();
+		}
+		String xml = createAuthenticationRequest(actual);
+		byte[] deflated = deflate(xml);
+		String encoded = encode(xml.getBytes(UTF_8));
+		String deflatedAndEncoded = encode(deflated);
+		List<Saml2X509Credential> signingCredentials = getSigningCredentials(request);
+		Map<String, String> qp = new LinkedHashMap<>();
+		addRelayState(request, qp);
+		if (request.isDeflate()) {
+			qp = signForDeflate(deflatedAndEncoded, signingCredentials, qp);
+		}
+		else {
+			qp.put("SAMLRequest", encoded);
+		}
+		MultiValueMap<String, String> result = new LinkedMultiValueMap<>();
+
+		for (Map.Entry<String, String> e : qp.entrySet()) {
+			result.add(e.getKey(), e.getValue());
+		}
+		return result;
+	}
+
+	private Map<String, String> signForDeflate(String deflatedAndEncoded,
+			List<Saml2X509Credential> signingCredentials,
+			Map<String, String> qp) {
+		//DEFLATE signature as query parameters
+		qp.put("SAMLRequest", deflatedAndEncoded);
+		qp = signingCredentials.isEmpty() ? qp : this.saml.signQueryParameters(signingCredentials, qp);
```

Comment:
This appears to be the only OpenSAML-specific line in this class. It makes me wonder 1. if the default implementation in the interface is missing something and 2. if this is the best place for the abstraction.

Reply:
OpenSAML specific items on [line 86](https://github.com/spring-projects/spring-security/blob/aeccdc562dd03a9a4d78d5e198144d92f469df0c/saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationRequestFactory.java#L86)

https://github.com/spring-projects/spring-security/blob/aeccdc562dd03a9a4d78d5e198144d92f469df0c/saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationRequestFactory.java#L86

Not sure how much of this could be extrapolated, or if we even want to.